data/crac/crac-impl/src/main/java/com/powsybl/openrao/data/cracimpl/ContingencyAdderImpl.java

```diff
@@ -34,31 +34,42 @@ protected String getTypeDescription() {
-        /*
-         * A contingency contains several network elements to trip. When adding a contingency, all the contained
-         * network elements have to be in the networkElements list of the crac.
-         * Here we go through all the network elements of the contingency, if an equal element is already present in
-         * the crac list we can directly pick its reference, if not we first have to create a new element of the
-         * list copying the network element contained in the contingency.
-         * Then we can create a new contingency referring to network elements already presents in the crac.
-         */
-        NetworkElement networkElement = this.owner.addNetworkElement(networkElementId, networkElementName);
-        this.networkElements.add(networkElement);
-        return this;
+    private void addElementsInContingency(ContingencyBuilder builder, String contingencyElementId, ContingencyElementType contingencyElementType) {
+        switch (contingencyElementType) {
+            case BRANCH -> builder.addBranch(contingencyElementId);
+            case GENERATOR -> builder.addGenerator(contingencyElementId);
+            case STATIC_VAR_COMPENSATOR -> builder.addStaticVarCompensator(contingencyElementId);
+            case SHUNT_COMPENSATOR -> builder.addShuntCompensator(contingencyElementId);
+            case HVDC_LINE -> builder.addHvdcLine(contingencyElementId);
+            case BUSBAR_SECTION -> builder.addBusbarSection(contingencyElementId);
+            case DANGLING_LINE -> builder.addDanglingLine(contingencyElementId);
+            case LINE -> builder.addLine(contingencyElementId);
+            case TWO_WINDINGS_TRANSFORMER -> builder.addTwoWindingsTransformer(contingencyElementId);
+            case THREE_WINDINGS_TRANSFORMER -> builder.addThreeWindingsTransformer(contingencyElementId);
+            case LOAD -> builder.addLoad(contingencyElementId);
+            case SWITCH -> builder.addSwitch(contingencyElementId);
+            case BATTERY -> builder.addBattery(contingencyElementId);
+            case BUS -> builder.addBus(contingencyElementId);
+            case TIE_LINE -> builder.addTieLine(contingencyElementId);
+        }
     }
 
     @Override
     public Contingency add() {
         checkId();
-        Contingency contingency = new ContingencyImpl(id, name, networkElements);
+        ContingencyBuilder builder = Contingency.builder(id);
+        elementsTypeById.forEach((id, type) -> addElementsInContingency(builder, id, type));
+        if (name != null) {
+            builder.addName(name);
+        }
+        Contingency contingency = builder.build();
         if (owner.getContingency(id) != null) {
-            if (owner.getContingency(id).equals(contingency)) {
+            if (owner.getContingency(id).equals(contingency) && owner.getContingency(id).getName().equals(contingency.getName())) {
```

Comment:
Name was added by me in the core for this issue, I was ask to add it as optional, make it parts of the equals so the identification of the object is a fundamental change I think, which was not the idea for an optional I think. In addition to that, Name was not part of previous rao contingency equals neither. We discuss briefly this issue before, but I added the check on Name to better detect some bug since I think contingency with same id but different Name are not expected in other parts of the code (do not remember exaclty where), but is is clearly a new check which was not existing before (I can remove it).

Reply:
It's better to leave it, but it's better still to have it inside the core Contingency IMO
@annetill what do you think?

Reply:
I remember why I added that: to make the message of the exception below true (before it was not true for name).

About changing equals in core, if I am not mistaken, adding Name in Contingency equals in core (or rao), will semantically means that 2 contingencies can have the same Id but different Name (including None vs something) and that should be considered ok and they will be treated as 2 differents objects contingencies, this is contradictory with all the time only the id of contingency is used to retrieve a contingency in core.

Reply:
I don't see why two contingencies should be able to co-exist with the same ID but different names (including one with, one without)

Reply:
I think there is a misunderstanding we should do a call about this.

Reply:
ok, a 10 minutes call would be enough I think, I'd like for @annetill to be present

Reply:
I think I understand what I misunderstood here:

I thought a contingency can be uniquely identified in core or in rao by using its Id, and if so, I was expecting the contingency equals to be:
```
    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Contingency other) {
            return id.equals(other.id);
        }
        return false;
    }
```
but is is not the case in core. (In my previous project it was using that kind of equals above (if any)). In core its seems that the equals are always "deep equals" meaning that we have to implement an equals and that equals need to be an equals on all attributs and children. If this is really the rule then we have to add Name since it is a new attribute. It could also be nice if we could add some kind of check that verify that equals is always a "deep equals" for everyone to know this core rule (also in case of inheritance etc it can be tricky).

The equals with all attributes will be something like:
```
    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Contingency other) {
            return id.equals(other.id) && name.equals(other.name) && elements.equals(other.elements);
        }
        return false;
    }
```

But none of the 1st or 2nd equals above verify in itself (expresses semantically) the rule that is on the message of rao adder:
`throw new OpenRaoException(format("A contingency with the same ID (%s) but a different name or network elements already exists.", this.id));`
which means that if 2 contingencies have the same id, they have to have the same name and children. Example of code:
```
        if (owner.getContingency(id) != null) { // here by definition owner.getContingency(id).getId() == contingency.getId()
            if (owner.getContingency(id).getElements().equals(contingency.getElements()) && owner.getContingency(id).getName().equals(contingency.getName())) {
                return owner.getContingency(id);
            } else {
                throw new OpenRaoException(format("A contingency with the same ID (%s) but a different name or network elements already exists.", this.id));
            }
        }
```

Reply:
given the 2nd equals is implemented in core, your code above is equivalent to:
~~~java
        if (owner.getContingency(id) != null) { // here by definition owner.getContingency(id).getId() == contingency.getId()
            if (owner.getContingency(id).equals(contingency)) {
                return owner.getContingency(id);
            } else {
                throw new OpenRaoException(format("A contingency with the same ID (%s) but a different name or network elements already exists.", this.id));
            }
        }
~~~

Reply:
As long as the core Contingency equals in not changed. For example, we can imagine that in core the equals is reduced to check only id for perf reason after an analysis that it should be unique by contingency. Then this check will no longer verify the expected property on name and elements.